### PR TITLE
Patch for setup.py egg_info issue

### LINKED ISSUE #5753: Release v2023.6.26: error: invalid command 'egg_info'
Upgrading to `pipenv` v2023.6.26 generates an `error: invalid command 'egg_info'` in the CircleCI builds of our Django project.

`pipenv` works as expected on local machines but fails on CircleCI until we `pip install --upgrade "pipenv<2023.6.26`

### Expected result

pipenv builds without errors.

### Actual result
```
Successfully installed pipenv-2023.6.26 setuptools-68.0.0
pipenv, version 2023.6.26
Creating a virtualenv for this project...
Pipfile: /home/circleci/circleci-workdir/Pipfile
Using /home/circleci/.pyenv/versions/3.9.17/bin/python3.9 (3.9.17) to create virtualenv...
created virtual environment CPython3.9.17.final.0-64 in 162ms
  creator CPython3Posix(dest=/home/circleci/circleci-workdir/.venv, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/home/circleci/.local/share/virtualenv)
    added seed packages: pip==23.1.2, setuptools==67.8.0, wheel==0.40.0
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator

Successfully created virtual environment!

Virtualenv location: /home/circleci/circleci-workdir/.venv
Installing dependencies from Pipfile.lock (ef2d1f)...
usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help

error: invalid command 'egg_info'

Exited with code exit status 1
```

When possible, provide the verbose output (`--verbose`), especially for locking and dependencies resolving issues.

### Steps to replicate

Provide the steps to replicate (which usually at least includes the commands and the Pipfile).

-------------------------------------------------------------------------------

Please run `$ pipenv --support`, and paste the results here. Don't put backticks (`` ` ``) around it! The output already contains Markdown formatting.

If you're on macOS, run the following:

    $ pipenv --support | pbcopy

If you're on Windows, run the following:

    > pipenv --support | clip

If you're on Linux, run the following:

    $ pipenv --support | xclip

### PR DESCRIPTION
**Pull request by matteius**
It was discovered that requirementslib uses the sys.executable python rather than the pipenv environment python.  Which as @oz123 reminded me should be the same, but due to some weird code paths, it wasn't always the same in my testing when the resolver got invoked as subprocess.

### The issue

Fixes #5753 

### The fix

I don't love the use of the environment variable, but I figured out this was the original intention of how the pipenv <--> requirements lib interface and at some point it got hard-coded to sys.executable in the resolver code which was getting invoked with a different python version at points.

This fixes the problem by ensuring that the correct python of the pipenv environment is being used.

### The checklist

* [X] Associated issue
* [ ] A news fragment in the `news/` directory to describe this fix with the extension `.bugfix.rst`, `.feature.rst`, `.behavior.rst`, `.doc.rst`. `.vendor.rst`. or `.trivial.rst` (this will appear in the release changelog). Use semantic line breaks and name the file after the issue number or the PR #.

### CODE REVIEW

File: pipenv/vendor/requirementslib/models/setup_info.py [hunk 1]
```diff
@@ -1141,29 +1137,15 @@ def run_setup(script_path, egg_base=None):
...
-        finally:
-            _setup_stop_after = None
-            sys.argv = save_argv
-            _setup_distribution = get_metadata(egg_base, metadata_type="egg")
-        dist = _setup_distribution
+        sp.run(
+            [python, "setup.py"] + args,
+            cwd=target_cwd,
+            stdout=sp.PIPE,
+            stderr=sp.PIPE,
```
Comment (cclauss) [inline pipenv/vendor/requirementslib/models/setup_info.py:1146]:
```suggestion
            capture_output=True
```
---
% `MY_CODE="import subprocess as sp ; sp.run([python, "setup.py"] + args, cwd=target_cwd, stdout=sp.PIPE, stderr=sp.PIPE)"`

% `echo $MY_CODE | ruff --select=UP022 --fix -`
```
import subprocess as sp ; sp.run([python, setup.py] + args, cwd=target_cwd, capture_output=True)
```
% `ruff rule UP022`
# replace-stdout-stderr (UP022)

Derived from the **pyupgrade** linter.

Autofix is always available.

## What it does
Checks for uses of `subprocess.run` that send `stdout` and `stderr` to a
pipe.

## Why is this bad?
As of Python 3.7, `subprocess.run` has a `capture_output` keyword argument
that can be set to `True` to capture `stdout` and `stderr` outputs. This is
equivalent to setting `stdout` and `stderr` to `subprocess.PIPE`, but is
more explicit and readable.

## Example
```python
import subprocess

subprocess.run(["foo"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

Use instead:
```python
import subprocess

subprocess.run(["foo"], capture_output=True)
```

## References
- [Python 3.7 release notes](https://docs.python.org/3/whatsnew/3.7.html#subprocess)
- [Python documentation: `subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run)

Comment (cclauss) [reply pipenv/vendor/requirementslib/models/setup_info.py]: Resolved in #5762

Comment (cclauss) [reply pipenv/vendor/requirementslib/models/setup_info.py]: Was the removal of `cwd=target_cwd` intentional?

Comment (matteius) [reply pipenv/vendor/requirementslib/models/setup_info.py]: @cclauss intentional, as it seemed unnecessary with the `cd(target_cwd)` context manager -- should already be in that directory.

Comment (matteius) [reply pipenv/vendor/requirementslib/models/setup_info.py]: Well its finally having a good run -- some just turned green.

**Comment by cclauss** [comment]
Thanks massively.  This fixed the problem for me.

**Pull request merged**